airbyte-cdk/python/airbyte_cdk/sources/streams/core.py

```diff
@@ -466,9 +467,7 @@ def _get_checkpoint_reader(
             return LegacyCursorBasedCheckpointReader(stream_slices=slices_iterable_copy, cursor=cursor, read_state_from_cursor=True)
         elif cursor:
             return CursorBasedCheckpointReader(
-                stream_slices=slices_iterable_copy,
-                cursor=cursor,
-                read_state_from_cursor=checkpoint_mode == CheckpointMode.RESUMABLE_FULL_REFRESH,
+                stream_slices=slices_iterable_copy, cursor=cursor, read_state_from_cursor=cursor.read_state_from_cursor()
```

Comment:
Before now, we had always operated under the assumption that if we were in RFR that the cursor should always read from the stream's state to determine if to continue syncing.

However, now that a substream can perform RFR by the checkpoint state of the single global parent state (incremental dependency), we'll end up infinitely looping when we look for the terminal state `{"__ab_full_refresh_sync_complete": True}` for each parent partition in state.

We need cursors to more explicitly delineate if a cursor should read state or just iterate. For now RFR specific cursors will return True. The others like GlobalStateCursor return False. The alternative is adding a type check, but I tend to hate inserting lots of one of type checks into our flow